File: fingerprintauthhelper/src/main/java/com/pro100svitlo/fingerprintAuthHelper/FahTimeOutService.java
package com.pro100svitlo.fingerprintAuthHelper;

import android.app.IntentService;
import android.content.Intent;
import android.os.Handler;

/**
 * Created by pro100svitlo on 8/29/16.
 */
public class FahTimeOutService extends IntentService {

    public final static String KEY_TRY_TIME_OUT = "KEY_TRY_TIME_OUT";
    public final static String TIME_OUT_BROADCAST = "TIME_OUT_BROADCAST";

    private Intent broadcastIntent;
    private final Handler mTimeOutHandler = new Handler();
    private long mTimeOutLeft = 0;
    private static boolean isRunning;

    public FahTimeOutService() {
        super(FahTimeOutService.class.getSimpleName());
    }

    public static boolean isRunning() {
        return isRunning;
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        isRunning = true;
        broadcastIntent = new Intent(TIME_OUT_BROADCAST);
        mTimeOutLeft = intent.getLongExtra(KEY_TRY_TIME_OUT, -1);
        timeoutRunnable.run();
    }

    private final Runnable timeoutRunnable = new Runnable() {
        @Override
        public void run() {
            isRunning = true;
            mTimeOutLeft = mTimeOutLeft - 1000;
            if (mTimeOutLeft > 0){
                mTimeOutHandler.postDelayed(timeoutRunnable, 1000);
            } else {
                isRunning = false;
                mTimeOutLeft = 0;
            }
            broadcastIntent.putExtra(FahManager.KEY_TIME_OUT_LEFT, mTimeOutLeft);
            sendBroadcast(broadcastIntent);
            if (!isRunning){
                stopSelf();
            }
        }
    };
}

File: fingerprintauthhelper/src/main/kotlin/com/pro100svitlo/fingerprintAuthHelper/FahTimeOutService.kt
package com.pro100svitlo.fingerprintAuthHelper

import android.app.IntentService
import android.content.Intent
import android.os.Handler

/**
 * Created by pro100svitlo on 11/23/16.
 */
class FahTimeOutService : IntentService("FahTimeOutService") {

    private val broadcastIntent: Intent by lazy { Intent(FahConstants.TimeOutService.TIME_OUT_BROADCAST) }
    private val mTimeOutHandler = Handler()

    companion object {
        private var timeOutLeft: Long = 0
        private var timeOut: Long = 0
        private var running: Boolean = false
        private var wasStoppedPreviously: Boolean = false

        fun isRunning(): Boolean {
            return running
        }

        fun tryToStopMe(): Boolean {
            if (wasStoppedPreviously) return false
            wasStoppedPreviously = true
            if (timeOut - timeOutLeft >= FahConstants.DEF_TRY_TIME_OUT) {
                timeOutLeft = 0
                return true
            } else {
                timeOutLeft = FahConstants.DEF_TRY_TIME_OUT - (timeOut - timeOutLeft)
            }
            return false
        }
    }

    override fun onHandleIntent(intent: Intent?) {
        running = true
        timeOut = intent?.getLongExtra(FahConstants.TimeOutService.KEY_TRY_TIME_OUT, -1) ?: -1
        timeOutLeft = timeOut
        if (timeOutLeft > 0) {
            timeoutRunnable.run()
        }
    }


    private var timeoutRunnable: Runnable = object :Runnable{
        override fun run() {
            running = true
            timeOutLeft -= 1000
            if (timeOutLeft > 0) {
                mTimeOutHandler.postDelayed(this, 1000)
            } else {
                running = false
                timeOutLeft = 0
            }
            broadcastIntent.putExtra(FahConstants.Manager.KEY_TIME_OUT_LEFT, timeOutLeft)
            sendBroadcast(broadcastIntent)
            if (!running) {
                stopSelf()
            }
        }
    }
}